### ISSUE TITLE
how to detect IP address assignment

### ISSUE THREAD

**Issue by evilgreen** [mwarning/zerotier-openwrt#66]
Is it possible programmatically detect the IP address assignment to ZeroTier device?
I was thinking about something similar to checking IFUPDATE_ADDRESSES variable available in _iface_ hotplug script. Currently the last ZeroTier event I can catch happens before the Internet connection is established which means that there is no IP on ztxxxxxxxx yet:

```
Mon Mar  9 20:48:29 2020 daemon.notice netifd: Interface 'ZT0' is enabled
Mon Mar  9 20:48:29 2020 daemon.notice netifd: Interface 'ZT0' is setting up now
Mon Mar  9 20:48:29 2020 daemon.notice netifd: Interface 'ZT0' is now up
Mon Mar  9 20:48:29 2020 daemon.notice netifd: Network device 'ztxxxxxxxx' link is up
Mon Mar  9 20:48:29 2020 daemon.notice netifd: Interface 'ZT0' has link connectivity
Mon Mar  9 20:48:29 2020 daemon.err zerotier-one[2209]: sendto: Network unreachable
Mon Mar  9 20:48:34 2020 user.notice URL-DEBUG: hotplug (iface): action='ifup' interface='ZT0'

```

Obviously one can run _zerotier-cli_ (or `ip -4 a show dev ...`) in a loop and wait for IP address to pop up, however that does not look nice.

**Comment by mwarning** [comment]
I would go with a hotplug script. Keep in mind that there are different hotplug directories in `/etc/hotplug.d`.
See https://openwrt.org/docs/guide-user/base-system/hotplug

**Comment by evilgreen** [comment]
Hotplug does not work here, at least for me, that is what I started from. As noted earlier, the last hotpug event happens much earlier than the Internet connection is established and IP address assigned to ztxxxxxxxx device. In other words, the 'ifup' below does not mean that the IP address is assigned:
`hotplug (iface): action='ifup' interface='ZT0'`

**Comment by mwarning** [comment]
ok, then I do not know what to do. Maybe ZT supports some kind of notifications...

**Comment by evilgreen** [comment]
It looks like the notification mechanism is available internally, see example in [5.1.4 Handling events](https://www.zerotier.com/manual/#5_1_4)
```
ZTS_EVENT_NETIF_UP --- network=a09acf023be465c1, mac=73b7abcfc207, mtu=10000
ZTS_EVENT_ADDR_NEW_IP4 --- addr=11.7.7.184 (on network=a09acf023be465c1)
ZTS_EVENT_ADDR_NEW_IP6 --- addr=fda0:9acf:233:e4b0:7099:9309:4c9b:c3c7 (on network=a09acf023be465c1)
ZTS_EVENT_NODE_ONLINE, node=c4c7ba3cf
ZTS_EVENT_NETWORK_READY_IP4 --- network=a09acf023be465c1
ZTS_EVENT_NETWORK_READY_IP6 --- network=a09acf023be465c1
```

Don't know how to use this though.

**Comment by mwarning** [comment]
I do not think we find a solution here. Maybe open a ticket at github.com/zerotier?
